vllm/v1/core/kv_cache_manager.py

```diff
@@ -92,6 +94,16 @@ def get_computed_blocks(self, request: Request) -> List[KVCacheBlock]:
                 hash_request_tokens(self.block_size, request))
         block_hashes = request.kv_block_hashes
 
+        if len(block_hashes) * self.block_size == request.num_tokens:
+            # When prompt length is divisible by the block size and all blocks
+            # are cached, we need to recompute the last token. This have to be
+            # achieved by re-computing an entire block because allocate_slots()
+            # assumes num_computed_tokens is always a multiple of the block
+            # size. This limitation can potentially be removed in the future to
+            # slightly improve the performance. To achieve this, the last block
+            # is removed from the computed block_hashes.
+            block_hashes = ConstantList(block_hashes[:-1])
```

Comment:
I was making a similar change but @WoosukKwon prefers to have it in scheduler, because this limitation is in the model runner instead of kv cache manager.

Reply:
Is there any suggestion on supporting sliding window? I don't want to make the complex special handling of sliding window in scheduler.

Reply:
What handling you plan to add for sliding window in scheduler? In general we should try to be modulization. For this particular logic, I think we could just

```python

class Scheduler:
    ...

    def maybe_recompute_last_block(self, computed_blocks, num_computed_tokens):
        ...

    def schedule(self):
        ...
        computed_blocks, num_computed_tokens = self.kv_cache_manager.get_computed_blocks(...)
        computed_blocks, num_computed_tokens = self.maybe_recompute_last_block(computed_blocks, num_computed_tokens)
```

Reply:
This interface is not very friendly to sliding window as the removal of last block needs a redo of `get_computed_blocks`. It works for me to discuss about which way is better after implementing the `maybe_recompute_last_block` for sliding window.
Is it ok to only change from ` computed_blocks = self.kv_cache_manager.get_computed_blocks(request)`  to ` computed_blocks, num_computed_tokens = self.kv_cache_manager.get_computed_blocks(request)` in this pr?

Reply:
Sure I'm ok with it.